### ISSUE TITLE
test with single image

### ISSUE THREAD

**Issue by irisadaf** [polimi-ispl/icpr2020dfdc#3]
I download weight files and I changed `test-model.py` to test pre-trained models with a single image, part of my codes is:

`device = 0
    net_name = 'EfficientNetAutoAttB4'
    net_class = getattr(fornet, net_name)
    # load model
    print('Loading model...')
    state_tmp = torch.load('model/ex.pth', map_location='cpu')
    if 'net' not in state_tmp.keys():
        state = OrderedDict({'net': OrderedDict()})
        [state['net'].update({'model.{}'.format(k): v}) for k, v in state_tmp.items()]
    else:
        state = state_tmp
    net: FeatureExtractor = net_class().eval().to(device)

    incomp_keys = net.load_state_dict(state['net'], strict=True)
    print(incomp_keys)
    print('Model loaded!')
    I = cv2.imread(img)
    I=cv2.resize(I,(224,224))
    I=cv2.cvtColor(I, cv2.COLOR_BGR2RGB)
    I=np.array(I,dtype=np.float)
    I=np.transpose(I)
    batch = torch.tensor(I/255).unsqueeze(0)
    with torch.no_grad():
            batch = batch.to(device, dtype=torch.float)
            output = net(batch)
            # get prediction
            score= output.cpu().numpy()[:, 0]`

I test with many real and fake images, but `score` is random (between -3 and 3) and didn't show anything about this image that is real or fake. what is the relation of score and class of image(real or fake)? Is my input image true? Is there any problem in my code?

**Comment by CrohnEngineer** [comment]
Hi @irisadaf ,

your code seems fine to me. Just two observations:

1. the output of the networks are not probabilistic, i.e. you have to apply a `sigmoid` function to the output of `net(batch)`  in order to have a value between 0 and 1 telling you the "fakeness" of your input (close to 1 = probably FAKE, close to 0 = probably REAL);
2. what kind of images are you testing? Are they still deepfakes? Are they deepfakes images taken from videos, or single images, etc...? Depending on the nature of your test set, I would probably suggest at least to finetune the networks on part of your dataset. Anyhow, without any other info I cannot be sure of what approach would suit best for you.

Hope this helps!
Cheers

Edoardo

**Comment by irisadaf** [comment]
@CrohnEngineer Thanks a lot
yes, they are deep fake images, My images are images of this dataset:
[https://www.kaggle.com/ciplab/real-and-fake-face-detection](https://www.kaggle.com/ciplab/real-and-fake-face-detection)

I apply the `sigmoid` function to the output of the network, but the score of 70% of cases are less than 0.5 and not detect fakeness of image.

**Comment by CrohnEngineer** [comment]
Hey @irisadaf ,

> Yes, they are deep fake images, My images are images of this dataset:
https://www.kaggle.com/ciplab/real-and-fake-face-detection

thank you for sharing this dataset with us, we didn't know about it :) 
If I have understood well, it is made by images realized with Photoshop only, meaning that the data you are testing the networks on is really different from the data on which they have been trained (which are frames extracted from automatically generated deepfake videos, with methods like [DeepFaceLab](https://github.com/iperov/DeepFaceLab)).
So, as I have written in the previous comment, I would suggest to finetune the networks on the training set provided by the authors. 

> I apply the sigmoid function to the output of the network, but the score of 70% of cases is less than 0.5 for both real and fake image folders!!!, and not detect fakeness of images. I tested them with EfficientNetAutoAttB4 model. The results are very sensitive to the input image size, for example, if I didn't resize images, the scores will be very low. The best results are for 224*224 size.

A couple of suggestions on that:

1. The size of the input images given to the network during training was 224x224 pixels, that's why you got the best results with these numbers. I would recommend to mantain this size also for finetuning;
2. Resizing is fine, but I would perform it only if the image size is smaller than 224x224 pixels; if the image is bigger, I would crop the region around the face of the subject having an image 224x224 pixels wide. You could use [Blazeface](https://github.com/hollance/BlazeFace-PyTorch) (or any other face detector/extractor you like) for doing that easily. 

Hope this helps with your work :) Have a nice weekend!
Cheers

Edoardo
